aptos-move/aptos-vm/src/verifier/transaction_arg_validation.rs

```diff
@@ -138,101 +205,154 @@ pub(crate) fn is_valid_txn_arg<S: MoveResolverExt>(
                 .expect("unreachable: struct must be allowed");
-            if let Some(validator) = option_validator {
-                validator(&s)?;
-            }
+            // By appending the BCS to the output parameter we construct the correct BCS format
+            // of the argument.
+            arg.append(&mut validate_and_construct(
+                session,
+                ty,
+                constructor,
+                allowed_structs,
+                cursor,
+                gas_meter,
+            )?);
         },
-        // this is unreachable given the check in `is_valid_txn_arg` and the
-        // fact we collect all arguments that involve strings and we validate
-        // them and them only
-        Bool | U8 | U16 | U32 | U64 | U128 | U256 | Address | Signer | Reference(_)
-        | MutableReference(_) | TyParam(_) => {
-            unreachable!("Validation is only for arguments with String")
+        Bool | U8 => read_n_bytes(1, cursor, arg)?,
+        U16 => read_n_bytes(2, cursor, arg)?,
+        U32 => read_n_bytes(4, cursor, arg)?,
+        U64 => read_n_bytes(8, cursor, arg)?,
+        U128 => read_n_bytes(16, cursor, arg)?,
+        U256 | Address => read_n_bytes(32, cursor, arg)?,
+        Signer | Reference(_) | MutableReference(_) | TyParam(_) => {
+            unreachable!("We already checked for this in is-valid-txn-arg")
         },
     };
 
     Ok(())
 }
 
+// A move function that constructs a type will return the BCS serialized representation of the
+// constructed value. This is the correct data to pass as the argument to a function taking
+// said struct as a parameter. In this function we execute the constructor constructing the
+// value and returning the BCS serialized representation.
+fn validate_and_construct<S: MoveResolverExt>(
+    session: &mut SessionExt<S>,
+    expected_type: &Type,
+    constructor: &FunctionId,
+    allowed_structs: &ConstructorMap,
+    cursor: &mut Cursor<&[u8]>,
+    gas_meter: &mut impl GasMeter,
+) -> Result<Vec<u8>, VMStatus> {
+    let (function, instantiation) = session.load_function_with_type_arg_inference(
+        &constructor.module_id,
+        constructor.func_name,
+        expected_type,
+    )?;
+    let mut args = vec![];
+    for param_type in &instantiation.parameters {
+        let mut arg = vec![];
+        recursively_construct_arg(
+            session,
+            &param_type.subst(&instantiation.type_arguments).unwrap(),
+            allowed_structs,
+            cursor,
+            gas_meter,
+            &mut arg,
+        )?;
+        args.push(arg);
+    }
+    let constructor_error = |e: VMError| {
+        // A slight hack, to prevent additional piping of the feature flag through all
+        // function calls. We know the feature is active when more structs then just strings are
+        // allowed.
+        let are_struct_constructors_enabled = allowed_structs.len() > 1;
+        if are_struct_constructors_enabled {
```

Comment:
Can we add a test case specifically for when struct constructors are disabled and we hit an error in utf8?

Reply:
This test case already exists in string_args